engine-jdk8-tests/src/test/java/org/hibernate/validator/test/internal/engine/valuehandling/OptionalTypeAnnotationConstraintTest.java

```diff
@@ -0,0 +1,216 @@
+
+import org.hibernate.validator.constraints.CompositionType;
+import org.hibernate.validator.constraints.ConstraintComposition;
+import org.hibernate.validator.constraints.NotBlank;
+import org.hibernate.validator.testutil.constraints.NotNullTypeUse;
+import org.hibernate.validator.valuehandling.UnwrapValidatedValue;
+
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import javax.validation.*;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Null;
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.lang.annotation.ElementType.*;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static org.hibernate.validator.testutil.ConstraintViolationAssert.*;
+import static org.hibernate.validator.testutil.ValidatorUtil.getValidator;
+
+/**
+ * @author Lukas Niemeier
+ */
+public class OptionalTypeAnnotationConstraintTest {
+
+	private Validator validator;
+
+	@BeforeClass
+	public void setup() {
+		validator = getValidator();
+	}
+
+	@Test
+	public void without_type_annotation_on_optional_is_validated_for_null_value() {
+		Model model = new Model();
+		model.valueWithoutTypeAnnotation = null;
+
+		Set<ConstraintViolation<Model>> constraintViolations = validator.validateProperty(
+				model,
+				"valueWithoutTypeAnnotation"
+		);
+		assertNumberOfViolations( constraintViolations, 1 );
+		assertCorrectPropertyPaths( constraintViolations, "valueWithoutTypeAnnotation" );
+		assertConstraintViolation( constraintViolations.iterator().next(), "container" );
+		assertCorrectConstraintTypes( constraintViolations, NotNull.class );
+	}
+
+	@Test
+	public void without_type_annotation_on_optional_is_validated_for_empty_value() {
+		Model model = new Model();
+		model.valueWithoutTypeAnnotation = Optional.empty();
+
+		Set<ConstraintViolation<Model>> constraintViolations = validator.validateProperty(
+				model,
+				"valueWithoutTypeAnnotation"
+		);
+		assertNumberOfViolations( constraintViolations, 0 );
+	}
+
+	/**
+	 * Actual: Fails with NullPointerException (Unexpected exception during isValid call)
+	 */
+	@Test
+	public void not_null_type_on_optional_is_validated_for_null_value() {
+		Model model = new Model();
+		model.valueWithNotNull = null;
+
+		Set<ConstraintViolation<Model>> constraintViolations = validator.validateProperty( model, "valueWithNotNull" );
+		assertNumberOfViolations( constraintViolations, 1 );
+		assertCorrectPropertyPaths( constraintViolations, "valueWithNotNull" );
+		assertConstraintViolation( constraintViolations.iterator().next(), "container" );
+		assertCorrectConstraintTypes( constraintViolations, NotNull.class );
+	}
+
+	/**
+	 * Actual: NotNull violation occurs on container's annotation
+	 */
+	@Test
+	public void not_null_type_on_optional_is_validated_for_empty_value() {
```

Comment:
This one makes me curious.

The property is annotated with `@UnwrapValidatedValue(false)`, so my personal expectation would be that the type parameter constraint is not applied at all (as it can only apply to the wrapped value, not the wrapper). That's different from the assertion in the test. I think I understand the reasoning between the expectation here, but it seems a bit incosistent to me.

@hferentschik, any thoughts?